### PR TITLE
drivers/si1145: initial implementation

### PR DESCRIPTION
**Pull request by aabadie**
Had this one in a pending branch. The proximity sensor is not implemented yet (well it still doesn't work) and I'm not sure about the UV, IR and visible light retrieved values.

### CODE REVIEW

**Comment by basilfx** [comment]
According to the datasheet, it is very similar to the Si1146 and Si1147. Would it make sense to group them (just like Si70xx)?

The difference: "With two or more LEDs, the Si1146/47 is capable of supporting multiple-axis proximity motion detection"

**Comment by aabadie** [comment]
> Would it make sense to group them (just like Si70xx)?

Yes the only difference is the address, on the I2C bus, IIRC.

**Comment by aabadie** [comment]
well and of course other internal configuration, as you mentioned.

File: drivers/include/si1145.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2016 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_si1145 SI1145
```
Comment (basilfx) [inline drivers/include/si1145.h]: I many be very nitpicky about the naming, but according to the datasheet it is Si1145 instead of SI1145 :-)

File: drivers/si1145/include/si1145_internals.h [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+#define SI1145_ALS_VIS_ALIGN         (1 << 4)
+#define SI1145_PS_RANGE              (1 << 5)
+#define SI1145_PS_ADC_MODE           (1 << 2)
+#define SI1145_VIS_RANGE             (1 << 5)
+#define SI1145_IR_RANGE              (1 << 5)
+#define SI1145_UCOEF0_DEFAULT        (0x7B)
+#define SI1145_UCOEF1_DEFAULT        (0x6B)
+#define SI1145_UCOEF2_DEFAULT        (0x01)
+#define SI1145_UCOEF3_DEFAULT        (0x00)
+/*#define SI1145_UCOEF0_DEFAULT        (0x29)
```
Comment (basilfx) [inline drivers/si1145/include/si1145_internals.h]: WIP I guess?

File: drivers/si1145/si1145.c [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+#include <string.h>
+
+#include "log.h"
+#include "si1145.h"
+#include "si1145_internals.h"
+#include "si1145_params.h"
+#include "periph/i2c.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG        (1)
```
Comment (basilfx) [inline drivers/si1145/si1145.c]: Debug is still toggled, still WIP I guess.

File: drivers/si1145/si1145.c [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+    /* Release I2C device */
+    i2c_release(dev->i2c_dev);
+
+    return 0;
+}
+
+void si1145_auto_init(void)
+{
+    for (unsigned i = 0; i < SI1145_NUMOF; i++) {
```
Comment (basilfx) [inline drivers/si1145/si1145.c]: At line 99, you use `int j`. Why only `unsigned i` here?

File: drivers/si1145/si1145.c [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+    /* set default UV measurement coefs */
+    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
+                  SI1145_REG_UCOEF0, SI1145_UCOEF0_DEFAULT);
+    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
+                  SI1145_REG_UCOEF1, SI1145_UCOEF1_DEFAULT);
+    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
+                  SI1145_REG_UCOEF2, SI1145_UCOEF2_DEFAULT);
+    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
+                   SI1145_REG_UCOEF3, SI1145_UCOEF3_DEFAULT);
```
Comment (basilfx) [inline drivers/si1145/si1145.c]: Space too much.

File: drivers/si1145/si1145_saul.c [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    si1145_read_distance(d, &distance);
+
+    res->unit = UNIT_M;
+    res->scale = 0;
+    return 1;
+}
+
+
+static int read_uv(void *dev, phydat_t *res)
```
Comment (basilfx) [inline drivers/si1145/si1145_saul.c]: What is the advantage of invoking another method with the exact same arguments. Compiler will optimize it out, but it increases LoC.

**Comment by basilfx** [comment]
Very nice sensor. I bought one right away on eBay. I can test it with ~2 weeks if needed :-)

File: drivers/include/si1145.h [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+extern const saul_driver_t si1145_uv_saul_driver;
+extern const saul_driver_t si1145_ir_saul_driver;
+extern const saul_driver_t si1145_visible_saul_driver;
+extern const saul_driver_t si1145_distance_saul_driver;
+/** @} */
+
+/**
+ * @brief auto-initialize all configured SI1145 devices
+ */
+void si1145_auto_init(void);
```
Comment (basilfx) [inline drivers/include/si1145.h]: So there are two ways for SAUL devices to auto init: either as part of `sys/auto_init/saul`, or in the driver itself. Not sure what is the preferred way?

I did it different for the Si70xx, which I found 'cleaner'.

File: drivers/si1145/si1145.c (hunk 4 above)
Comment (aabadie) [inline drivers/si1145/si1145.c]: probably because I started from the bmp180 implementation ;)

**Comment by aabadie** [comment]
@basilfx, thanks for the comments, I'll address them ASAP.

File: drivers/include/si1145.h (hunk 1 above)
Comment (aabadie) [inline drivers/include/si1145.h]: makes sense

File: drivers/si1145/include/si1145_internals.h (hunk 2 above)
Comment (aabadie) [inline drivers/si1145/include/si1145_internals.h]: yes, I found other implementations [here](https://github.com/adafruit/Adafruit_SI1145_Library) with different coefficients than the ones given in the datasheet. I wanted to try.

I guess you noticed it's WIP ;)

File: drivers/include/si1145.h (hunk 7 above)
Comment (aabadie) [inline drivers/include/si1145.h]: Not sure if it's cleaner for the Si70xx: code related to a specific device can be found in 2 different places. I prefer to have it all in a single location drivers/<device>. Maybe @haukepetersen can give his opinion on that (he's the father of saul) ?

Comment (basilfx) [inline drivers/include/si1145.h]: I guess one difference is that in `sys/auto_init/saul` the device is explicitly auto-initialized for SAUL, whereas this implementation auto initializes the device and (if desired) initializes it for SAUL.

File: drivers/include/si1145.h (hunk 1 above)
Comment (aabadie) [inline drivers/include/si1145.h]: I changed this one + quite a few others ;)

File: drivers/si1145/si1145.c (hunk 3 above)
Comment (aabadie) [inline drivers/si1145/si1145.c]: indeed, I'll remove this once the distance sensor works.

File: drivers/si1145/si1145.c (hunk 4 above)
Comment (aabadie) [inline drivers/si1145/si1145.c]: fixed

File: drivers/si1145/si1145.c (hunk 5 above)
Comment (aabadie) [inline drivers/si1145/si1145.c]: good catch! fixed

**Comment by basilfx** [comment]
Just received this sensor (China air mail was fast this time).

Hooked it up, and it seems to work. However, I doubt the return values (conditions did not change). Any idea?

```
+-------------------------------------+
RET code: 1 [NO ERROR]
RET code: 0 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
UV index: 5
IR light [lx]: 3073
Visible light [lx]: 1793
Distance [m]: 0

+-------------------------------------+
RET code: 1 [NO ERROR]
RET code: 0 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
UV index: 10
IR light [lx]: 4353
Visible light [lx]: 1
Distance [m]: 0

+-------------------------------------+
RET code: 1 [NO ERROR]
RET code: 0 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
UV index: 0
IR light [lx]: 3073
Visible light [lx]: 62208
Distance [m]: 0
```

File: tests/driver_si1145/main.c [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    si1145_t dev;
+    uint16_t uv = 0, ir = 0, visible = 0, distance = 0;
+    int result;
+
+    puts("SI1145 test application\n");
+
+    printf("+------------Initializing------------+\n");
+    result = si1145_init(&dev, TEST_I2C);
+    if (result == -1) {
+        puts("[Error] The given i2c is not enabled");
```
Comment (basilfx) [inline tests/driver_si1145/main.c]: Actually, the sensor was not found :-)

I would say: [Error] Unable to communicate with sensor.

**Comment by basilfx** [comment]
If you can give this PR a rebase, I can look at this again :-)

**Comment by aabadie** [comment]
@basilfx, rebased.

**Comment by basilfx** [comment]
I did a more in-depth attempt to get this to work, but I think some parts of the driver are not working or incomplete. I still get sensor values that make no sense to me. Any ideas? I tried to lower the I2C speed, increased sleeping times and double checked pull-up resistors.

With the sensor in complete dark

```
+-------------------------------------+
RET code: 1 [NO ERROR]
RET code: 0 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
UV index: 5
IR light [lx]: 65024
Visible light [lx]: 1537
Distance [m]: 1025

+-------------------------------------+
RET code: 1 [NO ERROR]
RET code: 0 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
UV index: 0
IR light [lx]: 513
Visible light [lx]: 2561
Distance [m]: 65280

+-------------------------------------+
RET code: 1 [NO ERROR]
RET code: 0 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
RET code: 1 [NO ERROR]
UV index: 2
IR light [lx]: 65280
Visible light [lx]: 65024
Distance [m]: 1
```

In the diff below, I changed the following (not related to above):

* Reverse the reset procedure (according to datasheet, `REG_HW_KEY` must be written first).
* Add LED current setup
* Force measurement when measuring proximity (verified with a red LED that now blinks).

```diff
diff --git a/drivers/si1145/si1145.c b/drivers/si1145/si1145.c
index 95ec334..efea5fc 100644
--- a/drivers/si1145/si1145.c
+++ b/drivers/si1145/si1145.c
@@ -193,6 +193,11 @@ int si1145_read_distance(si1145_t *dev, uint16_t *distance)
     /* Acquire exclusive access */
     i2c_acquire(dev->i2c_dev);

+    /* switch to force mode */
+    _set_als_force_mode(dev, SI1145_EN_PS1);
+
+    xtimer_usleep(285);
+
     /* read sensor here */
     char buffer[2] = { 0 };

@@ -218,6 +223,12 @@ int si1145_read_distance(si1145_t *dev, uint16_t *distance)
 static void _reset(si1145_t *dev)
 {
     DEBUG("Resetting sensor.\n");
+
+    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
+                  SI1145_REG_HW_KEY, SI1145_INIT_VALUE);
+
+    xtimer_usleep(WAIT_10MS);
+
     i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
                    SI1145_REG_MEAS_RATE0, 0);
     i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
@@ -226,11 +237,6 @@ static void _reset(si1145_t *dev)
     _send_command(dev, SI1145_RESET);

     xtimer_usleep(WAIT_10MS);
-
-    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
-                  SI1145_REG_HW_KEY, SI1145_INIT_VALUE);
-
-    xtimer_usleep(WAIT_10MS);
 }


@@ -247,6 +253,10 @@ static void _init_sensor(si1145_t *dev)
                   SI1145_REG_UCOEF2, SI1145_UCOEF2_DEFAULT);
     i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
                   SI1145_REG_UCOEF3, SI1145_UCOEF3_DEFAULT);
+
+    /* set LED1 current (see datasheet table 2) */
+    i2c_write_reg(dev->i2c_dev, SI1145_ADDR,
+                  SI1145_REG_PS_LED21, SI1145_LED1_CURRENT);
 }
```

**Comment by OlegHahm** [comment]
@aabadie, what's the state?

**Comment by aabadie** [comment]
> what's the state?

still WIP...

**Comment by aabadie** [comment]
I made a few changes on this one. It seems to work better. The proximity sensor needs an extra IR LED that I don't have.

I also plan to add saul integration. Can someone test again ?

**Comment by basilfx** [comment]
@aabadie Could you give it a rebase? I have the time (again) to look at this.

**Comment by aabadie** [comment]
@basilfx rebased. I didn't test this one for a long time but AFAIR it was not working. Feel free to take over and push directly to the PR's branch.

File: drivers/Makefile.dep [hunk 9]
```diff
@@ -301,8 +307,3 @@ ifneq (,$(filter xbee,$(USEMODULE)))
   USEMODULE += xtimer
   USEMODULE += netif
 endif
-
-ifneq (,$(filter lis3mdl,$(USEMODULE)))
```
Comment (smlng) [inline drivers/Makefile.dep]: unrelated?

Comment (smlng) [reply drivers/Makefile.dep]: and also: why?

File: tests/driver_si114x/main.c [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <stdio.h>
+#include <inttypes.h>
+
+#include "si114x.h"
+#include "si114x_params.h"
+#include "xtimer.h"
+#include "board.h"
+
+#define SLEEP_2S   (2 * 1000 * 1000u) /* 2 seconds delay between printf */
```
Comment (smlng) [inline tests/driver_si114x/main.c]: either use `US_PER_SEC` here or plain `xtimer_sleep` below.

File: tests/driver_si114x/main.c [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    result = si114x_init(&dev, &si114x_params[0]);
+    if (result == -SI114X_ERR_I2C) {
+        puts("[Error] The given i2c is not enabled");
+        return -1;
+    }
+    else if (result == -SI114X_ERR_NODEV) {
+        puts("[Error] The sensor did not answer correctly on the given address");
+        return 1;
+    }
+    else {
```
Comment (smlng) [inline tests/driver_si114x/main.c]: could omit `else` branch here, because of returns above.

File: drivers/si114x/si114x.c [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+#include <string.h>
+#include <stdint.h>
+
+#include "log.h"
+#include "si114x.h"
+#include "si114x_internals.h"
+#include "periph/i2c.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG        (1)
```
Comment (smlng) [inline drivers/si114x/si114x.c]: should be `0` before merge

File: drivers/si114x/include/si114x_internals.h [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+#define SI1147_ID                                                       (0x47)
+#define SI114X_STARTUP_TIME                                             (25000UL) /* 25ms */
+#define SI114X_WAIT_10MS                                                (10000UL) /* 10ms */
+#define SI114X_INIT_VALUE                                               (0x17)
+#define SI114X_UCOEF0_DEFAULT                                           (0x29)
+#define SI114X_UCOEF1_DEFAULT                                           (0x89)
+#define SI114X_UCOEF2_DEFAULT                                           (0x02)
+#define SI114X_UCOEF3_DEFAULT                                           (0x00)
+#define SI114X_ADC_REC_CLK                                              (0x70) /* 511 ADC Clock */
+
```
Comment (smlng) [inline drivers/si114x/include/si114x_internals.h]: missing group closing `/** @} */`

File: drivers/si114x/include/si114x_internals.h [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+#ifndef SI114X_INTERNALS_H
+#define SI114X_INTERNALS_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name Si114x I2C address
```
Comment (smlng) [inline drivers/si114x/include/si114x_internals.h]: not need to group a single define, right 😉

**Comment by smlng** [comment]
according to the discussion this won't make it into release 2018.01 right? Removing milestone, feel free to re-add.

File: drivers/Makefile.dep (hunk 9 above)
Comment (aabadie) [reply drivers/Makefile.dep]: A fail to resolve a conflict...

**Comment by basilfx** [comment]
@aabadie I have rebased your PR and made it compile again, but I cannot push them to your branch. (you can find it [here](https://github.com/basilfx/RIOT/tree/drivers_si1145)).

**Comment by basilfx** [comment]
Current state of this driver is that it is up-to-date with master, and compiles just fine. I've tested the driver, but I cannot get distance measuring to work, yet.

Will look into this, but once that works I think this driver is ready.

**Comment by aabadie** [comment]
> Will look into this, but once that works I think this driver is ready.

Thanks a lot for your help @basilfx !

**Comment by basilfx** [comment]
@aabadie I've updated and rebased, and I think it is ready for final polishing if you want.

One (major) issue that caused the ['strange' numbers](https://github.com/RIOT-OS/RIOT/pull/5667#issuecomment-235648526), was due to this:

```c
return (uint16_t)((buffer[0] << 8) | buffer[1]);
```

which should have been (and now is)

```c
return (uint16_t)((buffer[1] << 8) | buffer[0]);
```

At least, that's how the numbers got close/similar to the [Arduino](https://github.com/adafruit/Adafruit_SI1145_Library) example I took as a reference.

**Comment by basilfx** [comment]
Fixed build error due to wrong include guard.

@aabadie Can you do some additional testing and cleaning (if needed)?

**Comment by aabadie** [comment]
> Can you do some additional testing and cleaning (if needed)?

Sure I will, but not before this thursday afternoon. BTW thanks for the help !

**Comment by aabadie** [comment]
@basilfx, tested the PR again: the UV, light measures works (compared to Arduino library) but the distance seems weird. By the way, I opened #8605 because the SAUL class of sensors where marked as 'unknown' when using the `saul` shell commands.

**Comment by basilfx** [comment]
I think the distance is not returned as an absolute distance, but as the raw ADC count. It also depends on the LED (direct vs indirect) and type of material that covers the sensor. 

In a real world situation, you would calibrate the values, I think.

File: drivers/si114x/si114x_saul.c [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    res->scale = 0;
+    return 1;
+}
+
+static int read_distance(const void *dev, phydat_t *res)
+{
+    si114x_t *d = (si114x_t *)dev;
+
+    res->val[0] = si114x_read_distance(d);
+    res->unit = UNIT_M;
```
Comment (aabadie) [inline drivers/si114x/si114x_saul.c]: @basilfx, so it should be `UNIT_NONE` here ?

File: sys/auto_init/saul/auto_init_si114x.c [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+ */
+static saul_reg_t saul_entries[SI114X_NUMOF * 4];
+
+/**
+ * @brief   Define the number of saul info
+ */
+#define SI114X_INFO_NUMOF    (sizeof(si114x_saul_reg_info) / sizeof(si114x_saul_reg_info[0]))
+
+/**
+ * @brief   Reference the driver structs
```
Comment (aabadie) [inline sys/auto_init/saul/auto_init_si114x.c]: Reminder: should be `@name`

File: drivers/si114x/si114x_saul.c (hunk 15 above)
Comment (basilfx) [reply drivers/si114x/si114x_saul.c]: Yes.

The datasheet doesn't mention this explicitly, but [AN498](https://www.silabs.com/documents/public/application-notes/AN498.pdf) does (page 50).

File: sys/auto_init/saul/auto_init_si114x.c (hunk 16 above)
Comment (basilfx) [reply sys/auto_init/saul/auto_init_si114x.c]: I'll fix this and the above comment tonight.

Comment (basilfx) [reply sys/auto_init/saul/auto_init_si114x.c]: Fixed.

File: drivers/si114x/si114x_saul.c (hunk 15 above)
Comment (basilfx) [reply drivers/si114x/si114x_saul.c]: Changed this and also adapted test program to mention `distance [cnt]` instead of `distance [m]`.

Comment (basilfx) [reply drivers/si114x/si114x_saul.c]: And in the driver header file.

File: sys/auto_init/saul/auto_init_si114x.c (hunk 16 above)
Comment (aabadie) [reply sys/auto_init/saul/auto_init_si114x.c]: Just had a look at your changes and that's not what I was expecting. By the way, I updated the branch (force pushed)

**Comment by aabadie** [comment]
@basilfx, I rebased to fix a conflict (now that #8541 is merged) and changed a few minor things I found: 
* fixed some typos
* extended a bit the test application readme

I think now we are good. If it's ok on your side, can you ACK ?

**Comment by aabadie** [comment]
@smlng, your comments were also addressed

**Comment by basilfx** [comment]
Gave it another test, and ACK from my side.

**Comment by aabadie** [comment]
@smlng comments addressed. If it's fine for you, I'll squash

**Comment by basilfx** [comment]
Ping @smlng

**Comment by smlng** [comment]
please squash

**Comment by aabadie** [comment]
squashed

**Comment by basilfx** [comment]
Comments fixed, Murdock Green. Go!
